### PR TITLE
Raunak/ci update

### PR DESCRIPTION
**Pull request by RnkSngh**
PR to add listing and formatting to our CI (changes to actual solidity code were just those automatically created using forge fmt and should only include changes to white space or converting single quotes to double quotes.

Right now I'm just using the standard settings that come with the forge formatter, but this can be configured in the foundry.toml [here](https://book.getfoundry.sh/reference/config/formatter). If anyone has strong opinions on whitespace, single vs double quotes etc - speak now or forever hold your peace!

I will set up solhint + forge fmt check tomorrow morning!

## Summary by CodeRabbit

- **New Features**
    - Enhanced IBC channel management with new smart contracts and interfaces.
    - Introduced error handling libraries and custom errors for improved reliability.
    - Implemented GitHub Actions for linting and formatting checks.

- **Improvements**
    - Restructured import statements for better readability.
    - Refined error handling for clearer feedback and execution flow.
    - Updated function signatures and logic in smart contracts for consistency.
    - Added utility contracts for gas audit and consensus state management.

- **Refactor**
    - Consolidated event parameters and modified visibility for better encapsulation.
    - Reorganized contract imports and adjusted spacing for improved code clarity.

### CODE REVIEW

**Comment by RnkSngh** [comment]
This PR adds CI that lints using Solhint and also checks formatting, and makes fixes to our repo. 

This mostly uses default formatting/linting configurations, with a few changes to make the foundry formatter and the solhint linter compatible. The changes can be seen under the [fmt] section of `foundry.toml` and the `.solhint.json` file in the root.

refer to here for the specific formatting configs in case you'd like to see what we're enforcing/allowing:
- [foundry](https://book.getfoundry.sh/reference/config/formatter)
- [solhint](https://github.com/protofire/solhint/blob/HEAD/docs/rules.md)

The code changes to adhere to the new linting/formatting that are included in this PR were mainly: 

- run forge fmt with the updated `foundry.toml` file
- change all revert strings to [custom error messages](https://soliditylang.org/blog/2021/04/21/custom-errors/) - we were sometimes using revert strings and sometimes using custom error messages so I standardized to custom error messages (since it's a solidity best practice anyways) 
- change the ordering of some methods in some contracts to match generally acceptedd solidity conventions (e.g. internal functions should always come after external functions, etc) 
- change return values for functions to be declared in function declaration (we were doing this in some parts of our code and not in others - I think this[ makes code more readable](https://ethereum.stackexchange.com/questions/34046/what-is-the-purpose-of-giving-name-to-return-variables-in-function-declaration) so I standardized to always follow this)

**Comment by RnkSngh** [comment]
I think we should merge this PR before we merge the [proxy](https://github.com/open-ibc/vibc-core-smart-contracts/tree/raunak/upgradeable-dispatcher) one since that will make the proxy PR much easier to read

File: contracts/core/Dispatcher.sol [hunk 1]
```diff
@@ -452,18 +314,16 @@
     function recvPacket(IbcPacketReceiver receiver, IbcPacket calldata packet, Ics23Proof calldata proof) external {
         // verify `receiver` is the intended packet destination
         if (!portIdAddressMatch(address(receiver), packet.dest.portId)) {
-            revert receiverNotIndtendedPacketDestination();
+            revert IBCErrors.receiverNotIndtendedPacketDestination();
```
[the remark this answers is not in the document]
Comment (dshiell) [reply contracts/core/Dispatcher.sol:317]: @RnkSngh mind fixing up this typo and then I think we can get it in.

[the remark this answers is not in the document]
Comment (RnkSngh) [reply contracts/core/Dispatcher.sol]: should be fixed now!

File: test/VirtualChain.sol [hunk 2]
```diff
@@ -85,27 +85,24 @@ contract VirtualChain is Test, IbcEventsEmitter {
...
 
     // Assign new channelIds to both ends of the channel
-    function assignChanelIds(
-        IbcChannelReceiver localEnd,
-        IbcChannelReceiver remoteEnd,
-        VirtualChain remoteChain
-    ) external {
+    function assignChanelIds(IbcChannelReceiver localEnd, IbcChannelReceiver remoteEnd, VirtualChain remoteChain)
+        external
+    {
```
[the remark this answers is not in the document]
Comment (RnkSngh) [reply test/VirtualChain.sol]: this should also be fixed now!

File: contracts/core/Dispatcher.sol [hunk 3]
```diff
@@ -16,87 +20,32 @@ import '../interfaces/ConsensusStateManager.sol';
  *     which can be relayed to a rollup module on the Polymerase chain
  */
 contract Dispatcher is IbcDispatcher, IbcEventsEmitter, Ownable, Ibc {
-    //
-    // fields
-    //
     // IBC_PortID = portPrefix + address (hex string without 0x prefix, case insensitive)
     string public portPrefix;
-    uint32 portPrefixLen;
+    uint32 public portPrefixLen;
```
[the remark this answers is not in the document]
Comment (RnkSngh) [reply contracts/core/Dispatcher.sol:25]: @coderabbitai I think we can address this in subsequent PRs

File: contracts/core/Dispatcher.sol [hunk 4]
```diff
@@ -239,7 +124,7 @@
         CounterParty calldata counterparty,
         Ics23Proof calldata proof
     ) external {
-        verifyConnectIbcChannelProof(local, connectionHops, ordering, isChanConfirm, counterparty, proof);
+        _verifyConnectIbcChannelProof(local, connectionHops, ordering, isChanConfirm, counterparty, proof);
```
[the remark this answers is not in the document]
Comment (RnkSngh) [reply contracts/core/Dispatcher.sol:127]: @coderabbitai I think we can address this in subsequent PRs

File: contracts/core/Dispatcher.sol [hunk 5]
```diff
@@ -518,44 +374,159 @@
...
-
     // TODO: remove below writeTimeoutPacket() function
     //       1. core SC is responsible to generate timeout packet
     //       2. user contract are not free to generate timeout with different criteria
-    //       3. [optional]: we may wish relayer to trigger timeout process, but in this case, below function won't do the job, as it doesn't have proofs.
-    //          There is no strong reason to do this, as relayer can always do the regular `recvPacket` flow, which will do proper timeout generation.
+    //       3. [optional]: we may wish relayer to trigger timeout process, but in this case, below function won't do
+    // the job, as it doesn't have proofs.
+    //          There is no strong reason to do this, as relayer can always do the regular `recvPacket` flow, which will
+    // do proper timeout generation.
```
[the remark this answers is not in the document]
Comment (RnkSngh) [reply contracts/core/Dispatcher.sol:383]: @coderabbitai I think we can address this in subsequent PRs

**Review by dshiell — APPROVED** [review]
great first pass!

File: .github/workflows/lint.yml [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+      - uses: actions/checkout@v3
+        with:
+          submodules: recursive
+      - uses: actions/setup-node@v3
+        with:
+          node-version: '16'
+      - uses: bahmutov/npm-install@v1
+        with:
+          install-command: npm install
+      - run: npm run lint
```
Comment (nicopernas) [inline .github/workflows/lint.yml:25]: Does this work even without a package.json file in the repo?

Comment (nicopernas) [reply .github/workflows/lint.yml:25]: Ok. I see you added one but do we want one?

Comment (RnkSngh) [reply .github/workflows/lint.yml:25]: yes; I think solhint is only available via npm/yarn so we need a package.json I believe

**Pull request merged**